### PR TITLE
refactor(experimental): a function to turn a transaction into a durable nonce transaction

### PR DESCRIPTION
**Pull request by steveluscher**
refactor(experimental): a function to turn a transaction into a durable nonce transaction

## Summary

A durable nonce transaction is one whose:

1. Lifetime constraint is that of a durable nonce
2. First instruction is an `AdvanceNonceAccount` instruction against the system program

Use this function to turn a transaction into a durable nonce transaction by adding those two things.

## Test Plan

```
cd packages/transactions/
pnpm test:unit:browser
pnpm test:unit:node
```

### CODE REVIEW

File: packages/transactions/src/unsigned-transaction.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+export function getUnsignedTransaction<TTransaction extends BaseTransaction>(
+    transaction: TTransaction | (TTransaction & ITransactionWithSignatures)
+): TTransaction | Omit<TTransaction & ITransactionWithSignatures, keyof ITransactionWithSignatures> {
+    if ('signatures' in transaction) {
+        // The implication of the lifetime constraint changing is that any existing signatures are invalid.
+        const {
+            signatures: _, // eslint-disable-line @typescript-eslint/no-unused-vars
+            ...unsignedTransaction
+        } = transaction;
+        return unsignedTransaction;
```
Comment (buffalojoec) [inline packages/transactions/src/unsigned-transaction.ts:13]: This one is for my own understanding: Through the act of making this a durable nonce transaction, we void any existing signatures? Bc they are only valid for a recent blockhash, right

File: packages/transactions/src/instructions.ts [hunk 2]
```diff
@@ -39,8 +18,10 @@ export function prependTransactionInstruction<TTransaction extends BaseTransacti
...
 ): TTransaction | Omit<TTransaction, keyof ITransactionWithSignatures> {
-    const nextInstructions = [instruction, ...transaction.instructions] as const;
-    const out = replaceInstructions(transaction, nextInstructions);
+    const out = {
+        ...getUnsignedTransaction(transaction),
+        instructions: [instruction, ...transaction.instructions],
+    };
     Object.freeze(out);
     return out;
 }
```
Comment (buffalojoec) [inline packages/transactions/src/instructions.ts:27]: Assuming you offer `prepend` and `append` to add the nonce ixs before or after a set of arbitrary instructions?

File: packages/transactions/src/durable-nonce.ts [hunk 3]
```diff
@@ -16,20 +20,137 @@ type AdvanceNonceAccountInstruction<
...
+}
+
+function isDurableNonceTransaction(
+    transaction: BaseTransaction | (BaseTransaction & IDurableNonceTransaction)
+): transaction is BaseTransaction & IDurableNonceTransaction {
+    return (
+        'lifetimeConstraint' in transaction &&
+        typeof transaction.lifetimeConstraint.nonce === 'string' &&
+        transaction.instructions[0] != null &&
+        isAdvanceNonceAccountInstruction(transaction.instructions[0])
```
Comment (buffalojoec) [inline packages/transactions/src/durable-nonce.ts:117]: I think this answers my one question.
This first instruction is always `AdvanceNonce` in a DNT ?

**Review by buffalojoec — APPROVED** [review]
Lgtm, I haven't worked with durable nonces so I asked a couple of questions for my own understanding

File: packages/transactions/src/unsigned-transaction.ts (hunk 1 above)
Comment (steveluscher) [reply packages/transactions/src/unsigned-transaction.ts:13]: Basically the part of any transaction that needs to be signed _includes_ the blockhash/nonce as _content_. A change in the blockhash/nonce changes the transaction's content. Consequently, any existing signatures are no longer valid.

File: packages/transactions/src/instructions.ts (hunk 2 above)
Comment (steveluscher) [reply packages/transactions/src/instructions.ts:27]: In order for a transaction to make use of a nonce:

1. The nonce has to be set in place of the ‘recent blockhash’
2. The first instruction in the transaction _has_ to be the `AdvanceNonceAccount` instruction

So prepend only. No appending allowed.

https://docs.solana.com/implemented-proposals/durable-tx-nonces#runtime-support

Comment (steveluscher) [reply packages/transactions/src/instructions.ts:27]: In any case, these lines weren't part of this PR.

File: packages/transactions/src/durable-nonce.ts (hunk 3 above)
Comment (steveluscher) [reply packages/transactions/src/durable-nonce.ts:117]: Always!

**Pull request merged**